src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,422 @@
+	{
+		appendStringInfo(splitQuery, "%d,", shardinfo->nodeId);
+	}
+	appendStringInfo(splitQuery, "%d], %s)", shardinfo->nodeId, quote_literal_cstr(
+						 shardSplitMode));
+
+	return splitQuery;
+}
+
+
+/*
+ * It creates a background job for citus_split_shard_by_split_points and executes it in background.
+ */
+void
+ExecuteSplitBackgroundJob(int64 jobId, ShardInfo shardinfo, Datum datum,
+						  char *shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+	splitQuery = GetShardSplitQuery(shardinfo, datum, shardSplitMode);
+
+	/* ereport(LOG, (errmsg(splitQuery->data))); */
+	int32 nodesInvolved[] = { shardinfo->nodeId };
+	Oid superUserId = CitusExtensionOwner();
+	BackgroundTask *task = ScheduleBackgroundTask(jobId, superUserId, splitQuery->data, 0,
+												  NULL, 1, nodesInvolved);
+}
+
+
+/*
+ * This function executes a query and then decides whether a shard is subjected for isolation or average hash 2 way split.
+ * If a tenant is found splitpoints for isolation is returned otherwise average hash value is returned.
+ */
+Datum
+citus_find_shard_split_points(PG_FUNCTION_ARGS)
+{
+	int64 shardId = PG_GETARG_INT64(0);
+	int64 shardGroupSize = PG_GETARG_INT64(1);
+	ereport(DEBUG4, errmsg("%ld", shardGroupSize));
+
+	/*Filtering Shards with total GroupSize greater than MaxShardSize*1024 i.e Size based Policy*/
+	if (shardGroupSize < MaxShardSize * 1024)
+	{
+		PG_RETURN_NULL();
+	}
+
+	/*Extracting all the shardinfo with the help of shardId*/
+	Oid tableId = RelationIdForShard(shardId);
+	char *distributionColumnName = ColumnToColumnName(tableId,
+													  (Node *) DistPartitionKeyOrError(
+														  tableId));
+	char *dataType = format_type_be(ColumnTypeIdForRelationColumnName(
+										tableId,
+										distributionColumnName));
+	char *shardName = get_rel_name(tableId);
+	AppendShardIdToName(&shardName, shardId);
+	ShardInterval *shardrange = LoadShardInterval(shardId);
+	int64 shardMinValue = shardrange->minValue;
+	int64 shardMaxValue = shardrange->maxValue;
+	char *tableName = generate_qualified_relation_name(tableId);
+	StringInfo CommonValueQuery = makeStringInfo();
+
+	/*
+	 * The inner query for extracting the tenant value having frequency > TenantFrequency is executed on
+	 * every shard of the table and outer query gives the shardid and tenant values as the
+	 * output. Tenant Frequency is a GUC here.
+	 */
+	appendStringInfo(CommonValueQuery,
+					 "SELECT shardid , unnest(result::%s[]) from run_command_on_shards(%s,$$SELECT array_agg(val)"
+					 " FROM pg_stats s , unnest(most_common_vals::text::%s[],most_common_freqs) as res(val,freq)"
+					 " WHERE tablename = %s AND attname = %s AND schemaname = %s AND freq > %lf $$)"
+					 " WHERE result <> '' AND shardid = %ld;",
+					 dataType, quote_literal_cstr(tableName),
+					 dataType,
+					 quote_literal_cstr(shardName),
+					 quote_literal_cstr(distributionColumnName),
+					 quote_literal_cstr(get_namespace_name(get_rel_namespace(
```

Comment:
would be nicer to use a schemaName variable to avoid the deep nesting and messy indentation

Reply:
used schemaName